### ISSUE TITLE
Crash when deleting a hook in the hook callback

### ISSUE THREAD

**Issue by cbayet** [unicorn-engine/unicorn#1127]
Hi,

The following sample will make crash unicorn:

```
#!/usr/bin/python

from unicorn import Uc, UC_HOOK_CODE,UC_ARCH_X86, UC_MODE_64
# from unicorn.x86_const import 

code = "\x90"*10
addr = 0x1000

mu = Uc(UC_ARCH_X86, UC_MODE_64)

mu.mem_map(0x1000, 0x1000)
mu.mem_write(addr, code)


class MyHook(object):
    def __init__(self, mu, addr):

        self.mu = mu
        self.hook = mu.hook_add(UC_HOOK_CODE, lambda *args: self.my_hook(*args), begin=addr, end=addr)

    def my_hook(self, *args):
        print("Inside hook ! Removing...")
        self.mu.hook_del(self.hook)
        print("Hook deleted !")


h = MyHook(mu, 0x1001)

print("Starting emulation...")
mu.emu_start(addr, addr+len(code))
print("Emulation done !")
```

The problem come from the `uc_hook_del`method (https://github.com/unicorn-engine/unicorn/blob/master/uc.c#L1148) that doesn't have any ref count check or lock check on the elements of the list `uc->hooks`.

The `list_remove` is done anyways, freeing the list element. The element is still accessed in the  hook loop in `helper_uc_tracecode`, and is dereferenced to fetch the next hook, inducing a Use After Free situation and making the emulator crash.

First, the refs counter on the hooks should probably be incremented before calling its callback and decremented after.

Then, I guess we should have a list for hooks to remove that is called at a safe moment, and that will not remove the hooks that still have refs.

**Comment by aquynh** [comment]
if you can fix, please PR

**Comment by cbayet** [comment]
Fixed in #1130
